std/src/main/scala/algebra/std/list.scala

```diff
@@ -11,15 +12,21 @@ trait ListInstances {
 }
 
 class ListOrder[A](implicit ev: Order[A]) extends Order[List[A]] {
-  def compare(x: List[A], y: List[A]): Int =
-    (x, y) match {
-      case (Nil, Nil) => 0
-      case (Nil, _) => -1
-      case (_, Nil) => 1
-      case (x :: xs, y :: ys) =>
-        val n = ev.compare(x, y)
-        if (n != 0) n else compare(xs, ys)
-    }
+  def compare(xs: List[A], ys: List[A]): Int = {
+    @tailrec def loop(xs: List[A], ys: List[A]): Int =
+      xs match {
+        case Nil =>
```

Comment:
Ack - I think I've been sniped. Last benchmark, I swear! Added another benchmark which ensures the tuple is allocated on the heap (`*TupleHeap`). I've also included the bytecode scalac generates for the `checkTuple` method, just to be sure scalac wasn't doing anything fancy (it isn't). So, it seems like the JVM is taking care of the tuple strategy handily, which is nice. Results:

```
Benchmark                                              Mode  Samples          Score          Error  Units
a.b.TupleMatchBenchmark.emptyListLeft2Match           thrpt       10  253467840.436 ± 12223892.822  ops/s
a.b.TupleMatchBenchmark.emptyListLeft2MatchFast       thrpt       10  323188324.807 ± 13942236.297  ops/s
a.b.TupleMatchBenchmark.emptyListLeft2MatchFaster     thrpt       10  308222316.257 ± 17156861.142  ops/s
a.b.TupleMatchBenchmark.emptyListLeftTuple            thrpt       10  290671450.799 ± 15305844.711  ops/s
a.b.TupleMatchBenchmark.emptyListLeftTupleHeap        thrpt       10  162238249.825 ±  5178063.946  ops/s
a.b.TupleMatchBenchmark.emptyListRight2Match          thrpt       10  220221734.727 ±  3555704.452  ops/s
a.b.TupleMatchBenchmark.emptyListRight2MatchFast      thrpt       10  236875354.510 ±  4604759.625  ops/s
a.b.TupleMatchBenchmark.emptyListRight2MatchFaster    thrpt       10  332236731.746 ±  4843149.704  ops/s
a.b.TupleMatchBenchmark.emptyListRightTuple           thrpt       10  305556965.402 ±  3343750.517  ops/s
a.b.TupleMatchBenchmark.emptyListRightTupleHeap       thrpt       10  162082837.550 ±  7015805.533  ops/s
a.b.TupleMatchBenchmark.emptyLists2Match              thrpt       10  403974677.391 ±  3741407.088  ops/s
a.b.TupleMatchBenchmark.emptyLists2MatchFast          thrpt       10  374838359.717 ± 20196373.923  ops/s
a.b.TupleMatchBenchmark.emptyLists2MatchFaster        thrpt       10  382551994.617 ± 19779196.751  ops/s
a.b.TupleMatchBenchmark.emptyListsTuple               thrpt       10  383050848.374 ± 24689381.633  ops/s
a.b.TupleMatchBenchmark.emptyListsTupleHeap           thrpt       10  156116575.641 ± 10240649.596  ops/s
a.b.TupleMatchBenchmark.nonEmptyLists2Match           thrpt       10  204851392.131 ± 13503321.337  ops/s
a.b.TupleMatchBenchmark.nonEmptyLists2MatchFast       thrpt       10  199059823.003 ± 15640454.931  ops/s
a.b.TupleMatchBenchmark.nonEmptyLists2MatchFaster     thrpt       10  209463692.158 ± 12135568.592  ops/s
a.b.TupleMatchBenchmark.nonEmptyListsTuple            thrpt       10  203637089.638 ±  5339662.574  ops/s
a.b.TupleMatchBenchmark.nonEmptyListsTupleHeap        thrpt       10  126132252.642 ±  4469449.502  ops/s
```

For refrence, here is the bytecode generated by Scalac:

```
private int checkTuple(scala.collection.immutable.List<java.lang.Object>, scala.collection.immutable.List<java.lang.Object>);
  Code:
     0: new           #46                 // class scala/Tuple2
     3: dup
     4: aload_1
     5: aload_2
     6: invokespecial #50                 // Method scala/Tuple2."<init>":(Ljava/lang/Object;Ljava/lang/Object;)V
     9: astore_3
    10: aload_3
    11: ifnull        60
    14: aload_3
    15: invokevirtual #58                 // Method scala/Tuple2._1:()Ljava/lang/Object;
    18: checkcast     #60                 // class scala/collection/immutable/List
    21: astore        4
    23: aload_3
    24: invokevirtual #63                 // Method scala/Tuple2._2:()Ljava/lang/Object;
    27: checkcast     #60                 // class scala/collection/immutable/List
    30: astore        5
    32: getstatic     #69                 // Field scala/collection/immutable/Nil$.MODULE$:Lscala/collection/immutable/Nil$;
    35: aload         4
    37: invokevirtual #73                 // Method java/lang/Object.equals:(Ljava/lang/Object;)Z
    40: ifeq          60
    43: getstatic     #69                 // Field scala/collection/immutable/Nil$.MODULE$:Lscala/collection/immutable/Nil$;
    46: aload         5
    48: invokevirtual #73                 // Method java/lang/Object.equals:(Ljava/lang/Object;)Z
    51: ifeq          60
    54: iconst_0
    55: istore        6
    57: goto          199
    60: aload_3
    61: ifnull        90
    64: aload_3
    65: invokevirtual #58                 // Method scala/Tuple2._1:()Ljava/lang/Object;
    68: checkcast     #60                 // class scala/collection/immutable/List
    71: astore        7
    73: getstatic     #69                 // Field scala/collection/immutable/Nil$.MODULE$:Lscala/collection/immutable/Nil$;
    76: aload         7
    78: invokevirtual #73                 // Method java/lang/Object.equals:(Ljava/lang/Object;)Z
    81: ifeq          90
    84: iconst_m1
    85: istore        6
    87: goto          199
    90: aload_3
    91: ifnull        120
    94: aload_3
    95: invokevirtual #63                 // Method scala/Tuple2._2:()Ljava/lang/Object;
    98: checkcast     #60                 // class scala/collection/immutable/List
   101: astore        8
   103: getstatic     #69                 // Field scala/collection/immutable/Nil$.MODULE$:Lscala/collection/immutable/Nil$;
   106: aload         8
   108: invokevirtual #73                 // Method java/lang/Object.equals:(Ljava/lang/Object;)Z
   111: ifeq          120
   114: iconst_1
   115: istore        6
   117: goto          199
   120: aload_3
   121: ifnull        202
   124: aload_3
   125: invokevirtual #58                 // Method scala/Tuple2._1:()Ljava/lang/Object;
   128: checkcast     #60                 // class scala/collection/immutable/List
   131: astore        9
   133: aload_3
   134: invokevirtual #63                 // Method scala/Tuple2._2:()Ljava/lang/Object;
   137: checkcast     #60                 // class scala/collection/immutable/List
   140: astore        10
   142: aload         9
   144: instanceof    #75                 // class scala/collection/immutable/$colon$colon
   147: ifeq          202
   150: aload         9
   152: checkcast     #75                 // class scala/collection/immutable/$colon$colon
   155: astore        11
   157: aload         11
   159: invokevirtual #78                 // Method scala/collection/immutable/$colon$colon.head:()Ljava/lang/Object;
   162: invokestatic  #84                 // Method scala/runtime/BoxesRunTime.unboxToInt:(Ljava/lang/Object;)I
   165: istore        12
   167: aload         10
   169: instanceof    #75                 // class scala/collection/immutable/$colon$colon
   172: ifeq          202
   175: aload         10
   177: checkcast     #75                 // class scala/collection/immutable/$colon$colon
   180: astore        13
   182: aload         13
   184: invokevirtual #78                 // Method scala/collection/immutable/$colon$colon.head:()Ljava/lang/Object;
   187: invokestatic  #84                 // Method scala/runtime/BoxesRunTime.unboxToInt:(Ljava/lang/Object;)I
   190: istore        14
   192: iload         12
   194: iload         14
   196: isub
   197: istore        6
   199: iload         6
   201: ireturn
   202: new           #86                 // class scala/MatchError
   205: dup
   206: aload_3
   207: invokespecial #89                 // Method scala/MatchError."<init>":(Ljava/lang/Object;)V
   210: athrow
```